runbooks/how-to-load-cloudwatch-historical-data.md

```diff
@@ -0,0 +1,200 @@
+    --where partition_0 = '2020_(01-12)'                      --  9m13s, 32.58GB
+    --where partition_0 = '2021_(01-06)'                      -- 14m14s, 54.34GB
+    --where partition_0 = '2021_(07-12)'                      --  7m40s, 28.03GB
+    --where partition_0 = '2022_(01-05)'                      -- 13m21s, 52.49GB
+    --where partition_0 = '2022_(06-0903)'
+    --and month(from_iso8601_timestamp("timestamp")) <= 6   --  9m20s, 92.13GB
+    --where partition_0 = '2022_(06-0903)'
+    --and month(from_iso8601_timestamp("timestamp")) in (7,8) -- 16m05s, 92.13GB
+    where partition_0 = '2022_(06-0903)'                      --  4m08s, 92.46GB
+      and month(from_iso8601_timestamp("timestamp")) >= 9
+      and json_extract_scalar(message,'$.mdc["http_access_response_header_X-Request-ID"]') not in (
+            select "mdc_http_access_response_header_x-request-id"
+            from "bfd-insights-bfd-prod".bfd_insights_bfd_prod_api_requests
+      )
+   ) select
+      cw_timestamp,
+      cw_id,
+      timestamp,
+      level,
+      thread,
+      logger,
+      message,
+      context,
+      stuff['bene_id'] as "bene_id",
+      stuff['database_query_bene_by_coverage_batch'] as "database_query_bene_by_coverage_batch",
+      ... <repeat for all mdc columns in order since this is positional>
+      from dataset
+   ```
+
+6. Load the `target` table from the `staging` table.
+   1. The `target` table definition resides in terraform and should match the `staging` table columns and ordering with the one
+      difference being that the `staging` table does not include the `year` and `month` partition columns. Verify that
+      the table structure is the same if not done already. 
+   2. Similar to the loading of the `staging` table, this must be done in batches to avoid the Athena 30 minute timeout.
+      The batches here can be larger than for the `staging` table but ~500 million is the limit.
+      ```sql
+      insert into "bfd_insights_bfd_prod_api_requests" 
+      select
+         cw_timestamp,
+         cw_id,
+         timestamp,
+         level,
+         thread,
+         logger,
+         message,
+         context,
+         "mdc_bene_id",
+         "mdc_database_query_bene_by_coverage_batch",
+         "mdc_jpa_query_eobs_by_bene_id_snf_record_count",
+         ... <all mdc columns here>
+         date_format(from_iso8601_timestamp("timestamp"), '%Y') as "year",
+         date_format(from_iso8601_timestamp("timestamp"), '%m') as "month"
+      from bfd_cw_export.api_requests_no_partitions
+      --where year(from_iso8601_timestamp("timestamp")) = 2019
+      --where year(from_iso8601_timestamp("timestamp")) = 2020
+      --where year(from_iso8601_timestamp("timestamp")) = 2021
+      --where year(from_iso8601_timestamp("timestamp")) = 2022
+        --and month(from_iso8601_timestamp("timestamp")) <= 8
+      where year(from_iso8601_timestamp("timestamp")) = 2022
+         and month(from_iso8601_timestamp("timestamp")) = 8
```

Comment:
We don't need it here because the staging table was only loaded with records that are not duplicated in the target table. That is probably not clear so I will make an update accordingly.

Reply:
That's true only if the staging table partitions never overlap.

If, for example, there was an interruption in log ingestion on 10/01/22 and we exported/loaded a bunch of data for that date, and then there was another such interruption at the end of the same month. Unless you wipe the staging table between loads (which you could do), then you might have a mix of already-ingested and not-ingested data within the same year/month.

Reply:
It's definitely an edge case, though. I think a comment would suffice.